crypto/src/main/java/net/consensys/cava/crypto/sodium/DiffieHelman.java

```diff
@@ -0,0 +1,446 @@
+     * Create a {@link KeyPair} from pair of keys.
+     *
+     * @param publicKey The bytes for the public key.
+     * @param secretKey The bytes for the secret key.
+     */
+    public KeyPair(PublicKey publicKey, SecretKey secretKey) {
+      this.publicKey = publicKey;
+      this.secretKey = secretKey;
+    }
+
+    /**
+     * Create a {@link KeyPair} from an array of secret key bytes.
+     *
+     * @param secretKey The secret key.
+     * @return A {@link KeyPair}.
+     */
+    public static KeyPair forSecretKey(SecretKey secretKey) {
+      checkArgument(secretKey.ptr != null, "SecretKey has been destroyed");
+      return Sodium.scalarMultBase(secretKey.ptr, SecretKey.length(), (ptr, len) -> {
+        int publicKeyLength = PublicKey.length();
+        if (len != publicKeyLength) {
+          throw new IllegalStateException(
+              "Public key length " + publicKeyLength + " is not same as generated key length " + len);
+        }
+        return new KeyPair(new PublicKey(ptr, publicKeyLength), secretKey);
+      });
+    }
+
+    /**
+     * Generate a new key using a random generator.
+     *
+     * @return A randomly generated key pair.
+     */
+    public static KeyPair random() {
+      return Sodium.randomBytes(SecretKey.length(), (ptr, len) -> forSecretKey(new SecretKey(ptr, len)));
+    }
+
+    /**
+     * @return The public key of the key pair.
+     */
+    public PublicKey publicKey() {
+      return publicKey;
+    }
+
+    /**
+     * @return The secret key of the key pair.
+     */
+    public SecretKey secretKey() {
+      return secretKey;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == this) {
+        return true;
+      }
+      if (!(obj instanceof KeyPair)) {
+        return false;
+      }
+      KeyPair other = (KeyPair) obj;
+      return this.publicKey.equals(other.publicKey) && this.secretKey.equals(other.secretKey);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(publicKey, secretKey);
+    }
+  }
+
+  /**
+   * A Diffie-Helman shared secret.
+   */
+  public static final class Secret implements Destroyable {
+    @Nullable
+    private Pointer ptr;
+    private final int length;
+
+    private Secret(Pointer ptr, int length) {
+      this.ptr = ptr;
+      this.length = length;
+    }
+
+    @Override
+    protected void finalize() {
+      destroy();
+    }
+
+    @Override
+    public void destroy() {
+      if (ptr != null) {
+        Pointer p = ptr;
+        ptr = null;
+        Sodium.sodium_free(p);
+      }
+    }
+
+    @Override
+    public boolean isDestroyed() {
+      return ptr == null;
+    }
+
+    /**
+     * Compute a shared {@link Secret} from a secret key and a public key.
+     *
+     * @param secretKey the user's secret key
+     * @param publicKey another user's public key
+     * @return A shared {@link Secret}.
+     */
+    public static Secret forKeys(SecretKey secretKey, PublicKey publicKey) {
```

Comment:
So I created a `Secret` class for the shared secret obtained from scalar multiplication, as this keeps the secret in native memory where it can't get dumped and can easily be destroyed.

However, the only useful thing one can do with a `Secret` is to convert to a `byte[]` array anyway, which negates the above. So is it still worth it?

Reply:
It might be worth, although I'd need to prove it, to create a `Bytes` class backed by a `Secret`.

I think in the interim this is worth it, let's see how to iterate on this design. Thanks for the patches!